fuse_optimizers/src/fixed_lag_smoother.cpp

```diff
@@ -0,0 +1,320 @@
+  }
+  else
+  {
+    for (const auto& sensor_model_name : ignition_sensors_)
+    {
+      if (sensor_models_.find(sensor_model_name) == sensor_models_.end())
+      {
+        ROS_WARN_STREAM("Sensor '" << sensor_model_name << "' is configured as an ignition sensor, but no sensor "
+                        "model with that name currently exists. This is likely a configuration error.");
+      }
+    }
+    // Sort the sensors for efficient lookups
+    std::sort(ignition_sensors_.begin(), ignition_sensors_.end());
+  }
+
+  // Start the optimization thread
+  optimization_thread_ = std::thread(&FixedLagSmoother::optimizationLoop, this);
+
+  // Configure a timer to trigger optimizations
+  optimize_timer_ = node_handle_.createTimer(
+    optimization_period_,
+    &FixedLagSmoother::optimizerTimerCallback,
+    this);
+}
+
+FixedLagSmoother::~FixedLagSmoother()
+{
+  // Wake up any sleeping threads
+  optimization_running_ = false;
+  optimization_requested_.notify_all();
+  // Wait for the threads to shutdown
+  if (optimization_thread_.joinable())
+  {
+    optimization_thread_.join();
+  }
+}
+
+void FixedLagSmoother::preprocessMarginalization(const fuse_core::Transaction& new_transaction)
+{
+  timestamp_tracking_.addNewTransaction(new_transaction);
+}
+
+std::vector<fuse_core::UUID> FixedLagSmoother::computeVariablesToMarginalize()
+{
+  auto current_stamp = timestamp_tracking_.currentStamp();
+  auto lag_stamp = ros::Time(0, 0);
+  if (current_stamp > ros::Time(0, 0) + lag_duration_)
+  {
+    lag_stamp = current_stamp - lag_duration_;
+  }
+  auto old_variables = std::vector<fuse_core::UUID>();
+  timestamp_tracking_.query(lag_stamp, std::back_inserter(old_variables));
+  return old_variables;
+}
+
+void FixedLagSmoother::postprocessMarginalization(const fuse_core::Transaction& marginal_transaction)
+{
+  timestamp_tracking_.addMarginalTransaction(marginal_transaction);
+}
+
+void FixedLagSmoother::optimizationLoop()
+{
+  auto exit_wait_condition = [this]()
+  {
+    return this->optimization_request_ || !this->optimization_running_ || !ros::ok();
+  };
+  // Optimize constraints until told to exit
+  auto marginal_transaction = fuse_core::Transaction();
+  while (ros::ok() && optimization_running_)
+  {
+    // Wait for the next signal to start the next optimization cycle
+    auto optimization_deadline = ros::Time(0, 0);
+    {
+      std::unique_lock<std::mutex> lock(optimization_requested_mutex_);
+      optimization_requested_.wait(lock, exit_wait_condition);
+      optimization_deadline = optimization_deadline_;
+    }
+    // If a shutdown is requested, exit now.
+    if (!optimization_running_ || !ros::ok())
+    {
+      break;
+    }
+    // Apply motion models
+    auto new_transactions = fuse_core::Transaction::make_shared();
+    processQueue(*new_transactions);
+    // Prepare for selecting the marginal variables
+    preprocessMarginalization(*new_transactions);
```

Comment:
Just a bit of confusion about plurals here. The `preprocessMarginalization` takes a single `Transaction` object. A single transaction can contain multiple updates to the graph, but it's still one transaction. The description also makes it seem as though it's taking many transactions, but that would imply something like `std::vector<Transaction>`. Looking over the comments, it seems like in many places, we refer to a `Transaction` object, but then reference many transactions.

I do not feel strongly about this at all, though, so keep it as-is if it's a problem to change it.

Reply:
Yeah....I don't know. `new_transactions` is a single Transaction object that results from merging multiple queued Transaction objects into a single object. Does that make it plural or singular? If singular makes more sense to you, it is fine by me.

Reply:
I don't really feel strongly about it. It's a "special" kind of object, because many transactions can be represented by a single transaction.